packages/core/capabilities/core-capabilities-server-internal/src/resolve_capabilities.ts

```diff
@@ -6,46 +6,92 @@
  * Side Public License, v 1.
  */
 
-import { cloneDeep } from 'lodash';
+import { cloneDeep, memoize, uniqueId } from 'lodash';
 import { withSpan } from '@kbn/apm-utils';
+import type { CapabilitiesSwitcher } from '@kbn/core-capabilities-server';
 import type { KibanaRequest } from '@kbn/core-http-server';
 import type { Capabilities } from '@kbn/core-capabilities-common';
-import type { CapabilitiesSwitcher } from '@kbn/core-capabilities-server';
+import type { SwitcherWithOptions, SwitcherWithId } from './types';
+import { pathsIntersect, splitIntoBuckets, convertBucketToSwitcher } from './resolve_helpers';
 
-export type CapabilitiesResolver = (
-  request: KibanaRequest,
-  applications: string[],
-  useDefaultCapabilities: boolean
-) => Promise<Capabilities>;
+export type CapabilitiesResolver = ({
+  request,
+  capabilityPath,
+  applications,
+  useDefaultCapabilities,
+}: {
+  request: KibanaRequest;
+  capabilityPath: string[];
+  applications: string[];
+  useDefaultCapabilities: boolean;
+}) => Promise<Capabilities>;
+
+type ForPathSwitcherResolver = (path: string, switchers: SwitcherWithId[]) => string[];
+type AggregatedSwitchersResolver = (capabilityPaths: string[]) => CapabilitiesSwitcher[];
+
+export const getCapabilitiesResolver = (
+  getCapabilities: () => Capabilities,
+  getSwitchers: () => SwitcherWithOptions[]
+): CapabilitiesResolver => {
+  let initialized = false;
+  let capabilities: Capabilities;
+  let switchers: Map<string, SwitcherWithId>;
+  // memoize is on the first argument only by default, which is what we want here
+  const getSwitcherForPath: ForPathSwitcherResolver = memoize(getSwitchersToUseForPath);
```

Comment:
Adding cache on the "which switcher should I apply for this path" logic.

Note that we're calling this API for less than 10 different paths, so the amount in cache is negligible (plus, the return of the function is just references to permanent objects)

Note bis: we're scoping the cached method within the resolver, but it's mostly for test isolation: in practice, there's only one call to `getCapabilitiesResolver` and one single resolver for the whole Kibana instance.